File: 42 Traping Rain water/main.cpp
#include<iostream>
#include<vector>
using namespace std;

class Main {
    public:
        int trapingRainWater(vector<int>& vec) {
            int left = 0, right = vec.size() - 1;
            int leftMax = vec[left], rightMax = vec[right];

            int water = 0;

            while ( left < right ) {
                if (leftMax < rightMax) {
                    left += 1;
                    leftMax = max(leftMax, vec[left]);
                    water += leftMax - vec[left];
                } else {
                    right -= 1;
                    rightMax = max(rightMax, vec[right]);
                    water += rightMax - vec[right];
                }
            }

            return water;
        }
};

int main() {


    Main obj;

    vector<int> vec = {0,1,0,2,1,0,1,3,2,1,2,1};

    int res = obj.trapingRainWater(vec);

    cout << res << endl;

    return 0;
}

File: 42 Traping Rain water/main.py
class Main:
    def traping_rain_water(self, arr):
        left = 0
        right = len(arr) - 1

        left_max = arr[left]
        right_max = arr[right]

        water = 0

        while left < right:
            if arr[left] < arr[right]:
                left += 1
                left_max = max(left_max, arr[left])
                water += left_max - arr[left]
            else:
                right -= 1
                right_max = max(right_max, arr[right])
                water += right_max - arr[right]

        return water

obj = Main()
arr = [0,1,0,2,1,0,1,3,2,1,2,1]
res = obj.traping_rain_water(arr)
print(res)